packages/xstate-react/src/useMachine.ts

```diff
@@ -114,141 +96,86 @@ export function useMachine<
   Interpreter<TContext, any, TEvent, TTypestate>['send'],
   Interpreter<TContext, any, TEvent, TTypestate>
 ] {
-  const machine = useConstant(() => {
-    return typeof getMachine === 'function' ? getMachine() : getMachine;
-  });
-
-  if (
-    process.env.NODE_ENV !== 'production' &&
-    typeof getMachine !== 'function'
-  ) {
-    const [initialMachine] = useState(machine);
-
-    if (machine !== initialMachine) {
-      console.warn(
-        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
-          'Please make sure that you pass the same Machine as argument each time.'
-      );
-    }
-  }
-
-  const {
-    context,
-    guards,
-    actions,
-    activities,
-    services,
-    delays,
-    state: rehydratedState,
-    ...interpreterOptions
-  } = options;
-
-  const [resolvedMachine, service] = useConstant<
-    [
-      StateNode<TContext, any, TEvent, TTypestate>,
-      Interpreter<TContext, any, TEvent, TTypestate>
-    ]
-  >(() => {
-    const machineConfig = {
-      context,
-      guards,
-      actions,
-      activities,
-      services,
-      delays
-    };
-    const machineWithConfig = machine.withConfig(machineConfig, {
-      ...machine.context,
-      ...context
-    } as TContext);
+  const service = useInterpret(getMachine, options);
 
-    return [
-      machineWithConfig,
-      interpret(machineWithConfig, { deferEvents: true, ...interpreterOptions })
-    ];
+  useEffect(() => {
+    service.subscribe((s) => {
+      // Only change the current state if:
+      // - the incoming state is the "live" initial state (since it might have new actors)
+      // - OR the incoming state actually changed.
+      //
+      // The "live" initial state will have .changed === undefined.
+      const initialStateChanged =
+        s.changed === undefined && Object.keys(s.children).length;
+
+      if (s.changed || initialStateChanged) {
```

Comment:
This actually isn't an issue because the first `state` from `[state, send] = useMachine(...)` is always the initial/restored state, and _not_ the subscribed state. Added a test to show this.